### DISCUSSION TITLE
Get latest commit hash from remote URL

### DISCUSSION THREAD

**Discussion by jakewilliami** [GitoxideLabs/gitoxide#930] [Q&A] [answered]
Is it possible to get the latest commit hash from a remote URL with `gitoxide`?  I was thinking of using perhaps `gix-traverse` or `gix-commitgraph`, but I can't find any code that allows me to construct the commit graph from a remote URL.  I can also see that `gix::Repository::head_id` might be useful, but I don't know how I call this with a `gix::Remote`.

For example, here's how this would be done using `git2`:
```rust
let url = format!("https://github.com/{}/{}.git", GITHUB_USER, GITHUB_REPO_NAME);
let mut repo = git2::Remote::create_detached(url).expect("Cannot create remote repo instance");
repo.connect(git2::Direction::Fetch).expect("Cannot connect to remote repo");

let remote_heads = repo.list().expect("Cannot get remote head list from remote repo");
let commit_map: HashMap<String, String> = remote_heads
    .iter()
    .map(|rh| (rh.name().to_string(), rh.oid().to_string()))
    .collect();

repo.disconnect().expect("Cannot disconnect to remote repo");

// get /refs/heads/{branch} or HEAD from commit_map
```

So this can easily be done using `gitoxide`'s `git2` interface, however I'm wondering if there is an easier way to do it?

**Comment by Byron** [comment]
It is, and what you are looking for seems to be what's implemented in `gix remote -n https://github.com/git/git refs`. The code [can be found here](https://github.com/Byron/gitoxide/blob/dbc6cbb4363c2532f81b0bd6e351c4577bb9e9a3/gitoxide-core/src/repository/remote.rs#L75-L85). I hope that helps.

**Reply by jakewilliami** [reply]
That is somewhat helpful.  However, how is the remote repository constructed?  It looks like it's just passed as a function parameter.

Also note that I only care about the latest commit, not all of them, in case that changes the compute time.

**Reply by Byron** [reply]
> However, how is the remote repository constructed? It looks like it's just passed as a function parameter.

`Repository` instances can be [`open`](https://docs.rs/gix/0.48.0/gix/fn.open.html)ed or [`discover`](https://docs.rs/gix/0.48.0/gix/fn.discover.html)ed.

> Also note that I only care about the latest commit, not all of them, in case that changes the compute time.

One can restrict the selection of references to observe the last known state from [by controlling a remote's refspecs](https://docs.rs/gix/0.48.0/gix/struct.Remote.html#method.replace_refspecs).

**Reply by jakewilliami** [reply]
> `Repository` instances can be [`open`](https://docs.rs/gix/0.48.0/gix/fn.open.html)ed or [`discover`](https://docs.rs/gix/0.48.0/gix/fn.discover.html)ed.

Sorry to be a pain, but I'm still having trouble with this.  To be clear, I don't have the repository locally.  However, I know the URL.  It looks like both `open` and `discover` expect a local path to the repository.  Is it possible to construct a repository given _only its remote URL_, such as `git2`'s [`Remote::create_detached`](https://docs.rs/git2/latest/git2/struct.Remote.html#method.create_detached)?

**Reply by Byron** [reply, marked as answer]
**The real answer is that even though one can query remote references, one can't do so without a 'real' `Repository` instance.**

----
Thanks so much for pointing this out! Clearly that's a hole in the API as currently there is no way to get a Remote without a `Repository`, nor is it possible to get an in-memory repository. Right now the only workaround would be to have a physical repo on disk, but that's probably not really a workaround anyway.

It seems that the only reasonable thing to do with a Remote without Repository is to list remote references, and that could be put into its own utility function. After all, `gitoxide`'s plumbing crates are made to operate with only the bare minimum of information that's needed to perform a particular job. There is a lot wrong with having entirely standalone interactions with remotes as surprising amounts of configuration may contribute to successful connections in the first place, but that's certainly to be solved when fleshing out such API. Thinking about it again, it's probably easier to allow opening an 'anonymous' repository that loads all the usual configuration, but has no effective access to any `.git` repository, so [`Repository::remote_at(<url>)`](https://docs.rs/gix/0.48.0/gix/struct.Repository.html#method.remote_at) can be used naturally if some care is taken to specifically support this usecase. Maybe in the end, the API for that would end up being similar to the one in `git2` :).

**Reply by jakewilliami** [reply]
Thanks heaps Sebastian, good to know!  Happy to help discover holes in the API, and looking forward to what this might become in the future.  Gitoxide is amazing, and I love the ecosystem of minimal crates for specific tasks.  Great work, and thanks again :)
